pkg/printers: Support base64 decode in kubectl get go-template

glb opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

Adds a `base64decode` function to templates in `kubectl` so that it's possible to extract secret data in plaintext instead of base64 without requiring a separate executable to do the decode.

Sample usage:

```sh
kubectl get secret SECRET -o go-template='{{ .data.KEY | base64decode }}'
```

**Which issue(s) this PR fixes**:
Fixes kubernetes/kubernetes#45293.

**Special notes for your reviewer**:

**Release note**:
```release-note
You can now use the `base64decode` function in kubectl go templates to decode base64-encoded data, for example `kubectl get secret SECRET -o go-template='{{ .data.KEY | base64decode }}'`.
```

glb committed:
pkg/printers: Support base64 decode in kubectl go-template

Adds a `base64decode` function to templates in `kubectl` so that
it's possible to extract secret data in plaintext instead of
base64 without requiring a separate executable.

Sample usage:

```sh
kubectl get secret SECRET -o go-template='{{ .data.KEY | base64decode }}'
```

glb commented:
/sig cli

jdumars commented:
/ok-to-test

glb commented:
There seems to have been a segmentation fault [during the tests](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/60755/pull-kubernetes-verify/80877/):

```text
W0305 23:18:56.315] build/gen-swagger-docs.sh: line 29:     9 Aborted                 (core dumped) ./gradle-2.5/bin/gradle gendocs --info
```

Not sure about the cause, and there doesn't seem to be an open issue tracking this. Maybe a retest will not encounter the problem?

/retest

glb commented:
What's the protocol for reporting the seg fault?

glb commented:
@ericchiang @derekwaynecarr do you know when you might be able to get a chance to look at this PR (a tiny change with a fair amount of test boilerplate and then a test)? Is there someone else I should ask to take a look? I've tried to get feedback on whether this is a reasonable solution through questions on the issue but there's been nothing for the past week.

liggitt commented on pkg/printers/template.go:
```diff
@@ -112,3 +116,11 @@ func (p *TemplatePrinter) safeExecute(w io.Writer, obj interface{}) error {
+func base64decode(v string) string {
+	data, err := base64.StdEncoding.DecodeString(v)
+	if err != nil {
+		return err.Error()
```
this isn't good… it means the output from an error is indistinguishable from success

liggitt replied:
Make this a two value return:

> Each function must have either a single return value, or two return values of which the second has type error. In that case, if the second (error) return value evaluates to non-nil during execution, execution terminates and Execute returns that error.

glb replied:
Thanks! Will do.

glb replied:
I also figured out how to trigger the error path in the test using an intentionally-broken [Marshaler](https://golang.org/pkg/encoding/json/#Marshaler).

liggitt commented on pkg/printers/template.go:
```diff
@@ -33,7 +34,10 @@ type TemplatePrinter struct {
-		Funcs(template.FuncMap{"exists": exists}).
+		Funcs(template.FuncMap{
+			"exists": exists,
+			"b64dec": base64decode,
```
Is b64dec a name we're happy with long term? I have a slight preference for being explicit, e.g. base64decode

liggitt replied:
Also, the fact that that's the name you picked for the actual function (which I hadn't noticed) makes me smile

glb replied:
:) Thanks @liggitt ! I picked `b64dec` for compatibility with [sprig](https://github.com/Masterminds/sprig) on the off chance that folks might want to expand the set of functions using that library.

glb replied:
(and I got used to sprig from using Helm, so thought it would be good to be consistent.)

If you'd prefer `base64decode` I'm not stuck on `b64dec`, just wanted to explain where the name came from.

ericchiang replied:
I'd prefer `base64decode` too.

glb replied:
OK. Sorry to be slightly stuck on this, but would it be a reasonable compromise to support both `base64decode` and `b64dec` to reduce frustration for folks who are also writing helm templates? I say this because I got here by writing a helm template that needed to print out some initial values for secret data for the user doing the install, and I'm sure that me three weeks from now will trip over the different name.

Feel free to say no; I realize that having multiple names for the same function is potentially confusing in a different direction.

liggitt commented on pkg/printers/template.go:
```diff
@@ -112,3 +116,11 @@ func (p *TemplatePrinter) safeExecute(w io.Writer, obj interface{}) error {
 }
+
+func base64decode(v string) string {
+	data, err := base64.StdEncoding.DecodeString(v)
```
Does go json encoding use StdEncoding for []byte serialization?

glb replied:
Yes. 
![screen shot 2018-03-06 at 7 28 46 pm](https://user-images.githubusercontent.com/372794/37066717-afcc9d70-2174-11e8-9b02-5b729fe8f72c.png)

ericchiang commented on pkg/printers/template_test.go:
```diff
@@ -0,0 +1,77 @@
+		} else {
+			err = p.PrintObj(test.obj, buffer)
+		}
+		if err != nil {
```
Because of

```
err = p.PrintObj(test.obj, buffer)
```

`err` can never be `nil`

Also please avoid `else if`. For example

```
if err != nil && err.Error() != test.expectErr {
    t.Errorf("[%s]expect:\n %v\n but got:\n %v\n", name, test.expectErr, err)
    continue
}
if test.expectErr != "" {
    t.Errorf("[%s]expect:\n error %v\n but got:\n no error\n", name, test.expectErr)
}
```

glb replied:
Thanks @ericchiang! I'm not sure about the first comment -- `p.PrintObj(test.obj, buffer)` will return `nil` if there is no error?

I'll try to apply the second comment tomorrow; I am having some difficulty unwinding the logic I've got into the structure you requested without breaking the tests.

glb replied:
Just had to walk away for a couple of minutes, tests are passing now with simplified structure. Thanks for the comment, this looks a lot clearer now!

glb commented:
`pull-kubernetes-kubemark-e2e-gce` appears to be broken at the moment: https://github.com/kubernetes/kubernetes/issues/60870

glb commented:
From the comments it looks like the issue might be fixed, trying the test again.
/test pull-kubernetes-kubemark-e2e-gce

glb commented:
@liggitt @ericchiang I think I've addressed all of the comments. Let me know what you want to do on the function name, I'd prefer to be consistent with what users will see in Helm but will change if if you want. If everything is OK I can rebase and squash the commits.

ericchiang commented on pkg/printers/template_test.go:
```diff
@@ -0,0 +1,85 @@
+		"test error path for base64 decoding": {
+			template:  "{{ .Data.username | b64dec }}",
+			obj:       &badlyMarshaledSecret{},
+			expectErr: "error executing template \"{{ .Data.username | b64dec }}\": template: output:1:20: executing \"output\" at <b64dec>: error calling b64dec: illegal base64 data at input byte 0",
```
This relies on explicit error strings from several packages in the Go std library, which may change depending on the Go version used to run the test. This needs to only rely on error strings we control (e.g. check for a substring in the error), or only test for the presence of an error.

glb replied:
OK. I'll see what I can come up with. Thanks!

ericchiang reviewed (commented):
Thanks for the test flow changes. It's a lot easier to read.

Couple final nits. other than that lgtm

We're in freeze right now, so we're going to wait until master opens up for 1.11 in couple weeks or so.

glb commented:
Thanks @ericchiang ! Sorry, I didn't realize about the freeze, I'm sure you are all extremely busy with actual stuff more people want. Once I hear back on my final (I promise) question about the function name and fix the error test case, should I leave the commit stream until later or squash it? Then once it's done, will everything happen automagically once master opens up again or should I set a reminder to come back and ping you?

Thanks again!

glb commented:
/test pull-kubernetes-e2e-gce

ericchiang commented:
Please squash your commits. Other than that lgtm!

glb commented:
@ericchiang all squashed, commit & PR text updated to reflect `base64decode` instead of the original `b64dec`, tests passed. Good to go?

ericchiang commented:
/lgtm

ericchiang commented:
cc @sig-cli-maintainers

glb commented:
/assign @smarterclayton

smarterclayton commented:
/approve

although we should write up what our API compatibility story and put a comment in the code around what it entails (does this need to be longer term than our standard kubectl guarantees).

glb commented:
@smarterclayton sorry, is this something you'd like me to add to this PR? I'm not sure what to add here.

aleks-mariusz commented:
sorry to bump this old PR but i wanted to mention fi anyone else finds this via google that the answer on [this stackoverflow article](https://stackoverflow.com/a/58117444/8220409) makes _really_ nice use of it:

```
kubectl get secret name-of-secret -o go-template='
{{range $k,$v := .data}}{{printf "%s: " $k}}{{if not $v}}{{$v}}{{else}}{{$v | base64decode}}{{end}}{{"\n"}}{{end}}'

## In your case it would output
# password: decoded_password
# username: doceded_username
```
